### PR TITLE
[NG] Datagrid Render Refactor

### LINKED ISSUE #1150: If datagrid has horizontal scroll bar, the paging info gets truncated
**Select one ...**  (check one with "x")
```
[x ] bug
[ ] feature request
[ ] enhancement
```

### Actual behavior
Hi All, I'm facing this issue in IE. If datagrid has horizontal scroll bar, the paging info gets truncated.. Note: This is happening only in IE., In Chrome and firefox., it works fine.. Please refer attachment for screenshot..
datagrid_truncation_ie11

### Environment details

* **Angular version:** 2.0.X
4.0.1

* **Clarity version:** 
0.9.7

* **OS and version:** 
windows 10

* **Browser:** [ IE 11 ]

### LINKED ISSUE #1217: When horizontal scrolling is necessary, vertical scroll should be always visible and footer should not use the horizontal scroll
[x] bug
[ ] feature request
[ ] enhancement

### Expected behavior
If a datagrid is too wide to fit in the screen a horizontal scroll bar should appear. This scroll bar should be above the footer/within the table, so that:
1. I can see the vertical scroll all the time (if there is one)
2. I can see the whole footer (it makes no sense to scroll the footer since it has almost no information)

### Actual behavior
Currently the horizontal scroll appears below the footer. Initially I cannot see if there is a vertical scroll, and to use it, I first must scroll to the right, then drag the vertical scroll, then drag the horizontal scroll back to the left, so I can see the first (and presumable most important) column. Also since I can't see neither the vertical scroll, nor the items count (which is on the right side of the footer), I don't even know if there are more items in the datagrid
The same issue would come up if there is paging - the controls for it are on the right side and therefore - hidden.

### LINKED ISSUE #1409: Datagrid loading spinner is not centered
**Select one ...**  (check one with "x")
```
[x ] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior
Datagrid loading spinner should be centered both with and without `clr-dg-action-bar` buttons.

### Actual behavior

The style for `.datagrid-spinner` does not take into account the space added for the action bar. 

### Reproduction of behavior
https://plnkr.co/edit/XAUNqV?p=preview

### Environment details

* **Angular version:** 2.0.X

* **Clarity version:** 
Clarity latest

* **OS and version:** 
n/a

### LINKED ISSUE #1546: Dropdowns in datagrid action bar are misaligned
**Select one ...**  (check one with "x")
```
[x] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior

A dropdown in the datagrid action bar should line up with other buttons.

### Actual behavior

The dropdown is pushed down 1px.

### Reproduction of behavior

* Include a link to the reproduction scenario you created by forking one of the Clarity Plunker Templates: https://embed.plnkr.co/WpEcxC/

### Environment details

* **Angular version:** 4.4

* **Clarity version:** 0.10.6

* **OS and version:** 

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

### LINKED ISSUE #1631: Underlying checkboxex are in front of Dropdown body
```
[x ] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior
Use Dropdown in Datagrid ActionBar.

### Actual behavior
Checkboxes for selecting items are in front of dropdown body

### Reproduction of behavior
https://plnkr.co/edit/3nl7U7yjPXW9cHwI2gTr?p=preview

### Environment details

* **Angular version:** 4.x.x

* **Clarity version:** 0.10.10

* **OS and version:** 

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

### LINKED ISSUE #1675: `clr-dg-row-detail` Components should expand to take the full width of the parent
**Select one ...**  (check one with "x")
```
[x] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior

The `clr-dg-row-detail` component should expand to take the full width of its parent (row) container. 

### Actual behavior
The `clr-dg-row-detail` component is only as wide as its content. 

See the datagrid under Custom Component Row Details in [this](https://plnkr.co/edit/uVsLLzqcgnjTc9SXUyVG?p=preview) plunker for a reproduction scenario. 

### Environment details

* **Angular version:** 4.x.x

* **Clarity version:** `^0.10.12`

* **OS and version:** All

* **Browser:** [all]

# Workaround
For now the user can fix this by adding `flex: 1 1 auto;`to the child thats not expanding to fill its container. 

See this plunker for the workaround: 
https://plnkr.co/edit/uVsLLzqcgnjTc9SXUyVG?p=preview

### PR DESCRIPTION
**Pull request by hippee-lee**
[NG] Datagrid Render Refactor

## DOM Upates
This change includes several changes to datagrid child component templates. The changes address issues, add a more flexible structure for additional features and support the new way for rendering in the datagrid. Highlights include
1. Better scrolling for content overflow and a fixed footer
2. A DOM structure that supports sticky containers for cells and headers
3. Containers needed for various size calculations (done during a rendering cycle)

## Rendering Updates
1. Updates rendering for wrapped cells, rows and columns
2. Updates calculations for autosizing (still some issues)
3. Updates templates for fixed elements and vertical/horizontal scrolling
4. Removed old rendering for body, table and associated tests
5. Switch ClrDatagridItems to use NgForOf instance for server side iterator

## Notes
1. Datagrid popover elements (filters, hide/show toggle and row level signpost/tooltips) will still need manual workarounds until smart popovers is developed (#1262)
2. Known issues are tracked in issue #2683

## Fixes
1. Fixes #1409 - Loading spinner not centered
2. Fixes #1217 - Footer should not scroll when overflow-x
3. Fixes #1150 - horizontal scrollbar covers paging info
4. Fixes #1675 - `clr-dg-row-detail` should expand to full width of parent
5. Fixes #1631 - row selection checkboxes are in front of a dropdown in the `clr-dg-action-bar`
6. Fixes #1546 - Dropdowns in actionbar are misaligned

Signed-off-by: Matt Hippely <mhippely@vmware.com>

### CODE REVIEW

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 1]
```diff
@@ -891,31 +842,156 @@
...
       }
-      .datagrid-foot-select input[type='checkbox'] + label {
-        &::before,
-        &::after {
-          top: 0.11111rem;
-        }
+    }
+  }
+
+  .datagrid-footer-description {
```
Comment (mathisscott) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: `.datagrid-foot-description` => `.datagrid-footer-description` breaking change?

File: src/clr-angular/data/datagrid/all.spec.ts [hunk 2]
```diff
@@ -35,35 +35,40 @@ import DatagridRowSpecs from './datagrid-row.spec';
...
-import DatagridTableRendererSpecs from './render/table-renderer.spec';
+import WrappedCellSpec from './wrapped-cell.spec';
+import WrappedColumnSpec from './wrapped-column.spec';
+import WrappedRowSpec from './wrapped-row.spec';
 
 describe('Datagrid', function() {
   addHelpers();
 
   describe('Providers', function() {
+    /**/
```
Comment (mathisscott) [inline src/clr-angular/data/datagrid/all.spec.ts]: Wut?

File: src/dev/src/app/vertical-nav/nested-menus/vertical-nav-nested-menus.ts [hunk 3]
```diff
@@ -16,6 +16,7 @@ export class VerticalNavNestedMenusDemo {
 
   constructor(public verticalNavCases: VerticalNavCases) {
     this.case = this.verticalNavCases.allNestedMenu;
+    console.log(this.case);
```
Comment (mathisscott) [inline src/dev/src/app/vertical-nav/nested-menus/vertical-nav-nested-menus.ts]: This is my log. There are many like it but this one is mine.

File: golden/clr-angular.d.ts [hunk 4]
```diff
@@ -219,24 +219,31 @@ export declare class ClrControlHelper {
 export declare class ClrDatagrid<T = any> implements AfterContentInit, AfterViewInit, OnDestroy {
     SELECTION_TYPE: typeof SelectionType;
     allSelected: boolean;
+    calculationRows: ViewContainerRef;
```
Comment (youdz) [inline golden/clr-angular.d.ts]: Could you go over the new properties and methods in the golden file, and check for each one if it should be private, or if it needs to be public but underscore-prefixed?

File: src/clr-angular/data/datagrid/all.spec.ts [hunk 5]
```diff
@@ -83,18 +88,19 @@ describe('Datagrid', function() {
     DatagridColumnToggleSpecs();
     DatagridColumnToggleButtonSpecs();
     DatagridHideableColumnDirectiveSpec();
+    WrappedCellSpec();
+    WrappedColumnSpec();
+    WrappedRowSpec();
+    // DatagridProjectionSpec();
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/all.spec.ts]: Should this be on or off?

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 6]
```diff
@@ -5,60 +5,64 @@
 @include exports('datagrid.clarity') {
   @include basic-table(
     '.datagrid',
-    '.datagrid-head',
+    '.datagrid-header',
```
[the remark this answers is not in the document]
Comment (gnomeontherun) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: These class names are a breaking change, so we should note this in release.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 7]
```diff
@@ -67,226 +71,272 @@
...
+
+      .action-item {
+        @include clr-getTypePropertiesForDomElement(dropdown_text, (color, font-size, letter-spacing));
+
+        background: transparent;
+        border: 0;
+        color: $clr-app-font-color-primary;
+        cursor: pointer;
+        display: block;
+        line-height: calc(#{clr-getTypePropertyValueForDomElement(dropdown_text, line-height)} - 1px);
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: I'll updated this as it doesn't seem to matter either way.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 8]
```diff
@@ -374,284 +422,199 @@
...
-          right: 100%;
-          width: 0;
-          height: 0;
-          margin-top: -1 * $clr-datagrid-action-arrow-size;
-          border-right: $clr-datagrid-action-arrow-size solid $clr-datagrid-popover-border-color;
-          border-top: $clr-datagrid-action-arrow-size solid transparent;
-          border-bottom: $clr-datagrid-action-arrow-size solid transparent;
+          height: calc(100% + #{$clr-datagrid-column-separator-expandby} - #{$clr-default-borderwidth});
+          width: $clr-rem-1px;
+          top: calc(-0.5 * #{$clr-datagrid-column-separator-expandby} + 1px);
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: I'll updated this as it doesn't seem to matter either way.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 9]
```diff
@@ -374,284 +422,199 @@
...
+          background-color: $clr-table-border-color;
         }
 
-        &::after {
-          content: '';
+        .datagrid-column-handle {
+          @include clr-no-styles-button();
+          display: block;
+          position: absolute;
+          width: calc(0.5rem + 1px);
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Will update

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 10]
```diff
@@ -374,284 +422,199 @@
...
-          }
-
-          &.active {
-            background: $clr-datagrid-row-hover;
-            color: $clr-black;
-          }
+      .datagrid-signpost-trigger .signpost {
+        margin: -0.3rem 0;
+        height: 1.03rem;
+        .signpost-trigger {
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Yeah, I think it is. This code overrides the default signpost trigger style only when it is in the Column to keep the triggers centered. There is also one for the rows on line 221.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 11]
```diff
@@ -374,284 +422,199 @@
...
+        padding: 0.5rem;
+        display: flex;
+        flex-flow: column nowrap;
+        align-items: center;
+        justify-content: center;
+        justify-content: center;
+        font-size: clr-getTypePropertyValueForDomElement(datagrid_placeholder, font-size);
+        color: $clr-dark-midtone-gray;
+
+        .datagrid-placeholder-image {
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Nope. Will move up a level.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 12]
```diff
@@ -863,24 +809,29 @@
...
+        }
       }
-      outline-offset: $clr-datagrid-compact-outline-offset;
     }
-    .datagrid-foot {
+
+    .datagrid-footer {
       height: 1rem;
       padding: 0 $clr-table-cellpadding;
       line-height: calc(1rem - 1px);
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: This seems ok too. I will update and verify with regression tests.

File: src/clr-angular/data/datagrid/all.spec.ts (hunk 2 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/all.spec.ts]: Just trying to say something without saying anything. 
j/k will remove.

File: src/dev/src/app/vertical-nav/nested-menus/vertical-nav-nested-menus.ts (hunk 3 above)
Comment (hippee-lee) [reply src/dev/src/app/vertical-nav/nested-menus/vertical-nav-nested-menus.ts]: Removed

File: src/clr-angular/data/datagrid/providers/table-size.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  DragDispatcher,
+  Page,
+  StateDebouncer,
+];
+
+export default function(): void {
+  describe('TableSizeService', function() {
+    beforeEach(function(this: TestContext) {
+      TestBed.configureTestingModule({
+        imports: [ClarityModule],
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/providers/table-size.service.spec.ts]: I'll update to only import the `ClrDatagridModule` for testing.

File: src/clr-angular/data/datagrid/providers/display-mode.service.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+import { DatagridRenderOrganizer } from '../render/render-organizer';
+
+@Injectable()
+export class DisplayModeService {
+  protected _view: BehaviorSubject<DatagridDisplayMode> = new BehaviorSubject<DatagridDisplayMode>(
+    DatagridDisplayMode.DISPLAY
+  );
+
+  constructor(renderOrganizer: DatagridRenderOrganizer) {
+    renderOrganizer
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/providers/display-mode.service.ts]: I didn't think this was needed for services. I thought they would get cleaned up when the component they were injected into was destroyed but it looks like you are right from the [OnDestroy Docs](https://angular.io/api/core/OnDestroy). 

Will remove.

File: src/clr-angular/data/datagrid/providers/display-mode.service.spec.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    it('exposes an Observable for display mode view state', function() {
+      const viewObservable = this.displayService.view;
+      expect(viewObservable).toBeDefined();
+      expect(viewObservable instanceof Observable).toBe(true);
+    });
+
+    it('properly updates the view mode when organizer resizes', function(this: UserContext) {
+      let currentChange: DatagridDisplayMode;
+      let displayChangeCount = 0;
+      this.displayService.view.subscribe(viewChange => {
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/providers/display-mode.service.spec.ts]: I added a field to the context to store and unsubscribe from the subscription.

File: src/clr-angular/data/datagrid/helpers.spec.ts [hunk 16]
```diff
@@ -83,6 +83,27 @@ export function addHelpers(): void {
...
+        imports: [ClarityModule],
+        declarations: [testComponent, ...extraDirectives],
+        providers: providers,
+      })
+        .overrideComponent(clarityDirective, {
+          set: {
+            providers: serviceOverrides,
+          },
+        })
+        .compileComponents();
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/helpers.spec.ts]: Correct. I will remove.

File: src/clr-angular/data/datagrid/datagrid.ts [hunk 17]
```diff
@@ -225,6 +243,39 @@ export class ClrDatagrid<T = any> implements AfterContentInit, AfterViewInit, On
         }
       })
     );
+    this.displayMode.view.subscribe(viewChange => {
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/datagrid.ts]: I added some comments but can refine if its not clear.

File: src/clr-angular/data/datagrid/datagrid-items.ts [hunk 18]
```diff
@@ -3,55 +3,68 @@
...
-    private _differs: IterableDiffers,
-    private _items: Items<T>
+    private differs: IterableDiffers,
+    private _items: Items,
+    private vcr: ViewContainerRef,
+    private _differs: IterableDiffers
   ) {
     _items.smartenUp();
+    this.iterableProxy = new NgForOf<T>(this.vcr, this.template, this.differs);
+    _items.change.subscribe(items => {
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/datagrid-items.ts]: Added an unsubscribe.

File: src/clr-angular/data/datagrid/datagrid-items.spec.ts [hunk 19]
```diff
@@ -14,73 +14,150 @@ import { Page } from './providers/page';
...
-      this.testComponent = this.fixture.componentInstance;
-      this.clarityDirective = this.fixture.componentInstance.datagridItems;
-      this.itemsProvider = TestBed.get(Items);
-    });
 
-    afterEach(function() {
-      this.fixture.destroy();
-    });
+      afterEach(function() {
+        this.fixture.destroy();
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/datagrid-items.spec.ts]: Removing this line doesn't seem to affect tests. I will remove.

File: src/clr-angular/data/datagrid/datagrid-items.spec.ts [hunk 20]
```diff
@@ -14,73 +14,150 @@ import { Page } from './providers/page';
...
+        this.testComponent = this.fixture.componentInstance;
+        this.clarityDirective = this.fixture.componentInstance.datagridItems;
+        this.itemsProvider = TestBed.get(Items);
+      });
 
-@Component({ template: `<div *clrDgItems="let n of numbers; trackBy: trackBy">{{n}}</div>` })
-class FullTest {
-  @ViewChild(ClrDatagridItems) datagridItems: ClrDatagridItems<number>;
+      afterEach(function() {
+        this.fixture.destroy();
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/datagrid-items.spec.ts]: Removing this line doesn't seem to affect tests. I will remove.

File: src/clr-angular/data/datagrid/all.spec.ts (hunk 5 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/all.spec.ts]: Merge artifact will remove as this suite was replaced by the three `Wrapped*Spec`'s
Might have ben a rebase artifact. Will remove.

File: src/clr-angular/data/datagrid/datagrid-items.spec.ts [hunk 21]
```diff
@@ -64,10 +64,6 @@ export default function(): void {
         this.itemsProvider = TestBed.get(Items);
       });
 
-      afterEach(function() {
-        this.fixture.destroy();
```
Comment (youdz) [inline src/clr-angular/data/datagrid/datagrid-items.spec.ts]: I think I missed something, why are you removing the `afterEach`?

File: src/clr-angular/data/datagrid/datagrid-row-detail.ts [hunk 22]
```diff
@@ -97,7 +110,12 @@ export class ClrDatagridRowDetail<T = any> implements AfterContentInit, OnDestro
     });
   }
 
+  // toggleExpand() {
```
Comment (youdz) [inline src/clr-angular/data/datagrid/datagrid-row-detail.ts]: You can just remove that if we don't use it. You introduced it in this topic branch, so it's not a breaking change. 😉

File: src/clr-angular/data/datagrid/datagrid-items.spec.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/datagrid-items.spec.ts]: I reverted this back to having the `afterEach`, it is needed for cleanup and will prevent any memory leaks in the tests.

File: src/clr-angular/data/datagrid/datagrid-items.spec.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/datagrid-items.spec.ts]: I reverted this back to having the `afterEach`, it is needed for cleanup and will prevent any memory leaks in the tests.

**Pull request merged**